text/0020-block-proposer-improvements.md

```diff
@@ -0,0 +1,108 @@
+
+However, the issue that can arise, is that Consensus issues a notification about finalization of a block, and a request for deploys shortly afterwards (so that it doesn't include the newly finalized block in the information about the past of the block to be proposed), but the Block Proposer receives these events in reverse order, thus making it not aware that the newly finalized block should also be considered a part of the new block's past.
+
+The solution for this issue could be to include the hash of the last finalized block in the request for deploys. The Block Proposer could then check if it already has that block as finalized, and if not, it would buffer the request and complete it when it receives the notification of the relevant block's finalization.
+
+Even with this solution in place, we could actually still experience problems. Consensus could finalize multiple blocks at once, and if the events aren't guaranteed to arrive in order, the Block Proposer could be notified of the finalization of the last block before some others. In such a case, it would still have an incomplete picture of the new block's past. To prevent this, we would have to only handle notifications about finalized blocks if the block's parent has already been finalized (and buffer finalization notifications that don't satisfy this condition, until it is satisfied).
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+### Seeing finalization before proposal
+
+When receiving `FinalizedProtoBlock(block)`:
+
+- check that `block.hash()` is in `self.state.proposed`
+- if not, add `block.hash()` to a new collection `self.state.finalization_queue`
+- if yes, proceed in the current manner
+
+When receiving `ProposedProtoBlock(block)`:
+
+- check if `block.hash()` is in `self.state.finalization_queue`
+- if yes, remove it from `finalization_queue` and proceed as if we received both `ProposedProtoBlock` and `FinalizedProtoBlock`
+- if no, proceed in the current manner
+
+### Request for a block before seeing another validator's proposal
+
+This would require adding a `last_finalized_block: BlockHash` field to `BlockProposerRequest::ListForInclusion`.
+
+When receiving a `ListForInclusion` request:
+
+- check if `last_finalized_block` is in `self.state.finalized`
+- if yes, proceed in the current manner
+- if not, buffer the request in a new field `self.state.request_queue`
+
+When receiving a `FinalizedProtoBlock` event:
+
+- check if the block's parent is in `self.state.finalized` (in addition to other checks that are in place)
```

Comment:
That means we'll have to store `BlockHash` instead of `ProtoBlock` in that case, does it? (`ProtoBlock` doesn't have an explicit parent field.)
Which is fine, I think, because `ProtoBlock` can be computed from it.